docs/DafnyRef/Grammar.md

```diff
@@ -271,163 +168,115 @@ method m() {
-    "object" | "object?" | "old" | "opened" | "ORDINAL"
-    "predicate" | "print" | "provides" |
-    "reads" | "real" | "refines" | "requires" | "return" |
-    "returns" | "reveal" | "reveals" |
-    "seq" | "set" | "static" | "string" |
-    "then" | "this" | "trait" | "true" | "twostate" | "type" |
-    "unchanged" | "var" | "while" | "witness" |
-    "yield" | "yields" |
-    arrayToken | bvToken
-
-arrayToken = "array" [ posDigitFrom2 | posDigit digit { digit }]["?"]
-
-bvToken = "bv" ( 0 | posDigit { digit } )
-````
-
-An ``arrayToken`` is a reserved word that denotes an array type of
-given rank. `array` is an array type of rank 1 (aka a vector). `array2`
-is the type of two-dimensional arrays, etc.
-`array1` and `array1?` are not reserved words; they are just ordinary identifiers.
-Similarly, `bv0`, `bv1`, and `bv8` are reserved words, but `bv02` is an
-ordinary identifier.
+
+Dafny has a set of reserved words that may not
+be used as identifiers of user-defined entities.
+These are listed [here](#sec-g-tokens).
+
+In particular note that
+
+- `array`, `array2, `array3` , etc. are reserved words, but not `array1` or `array0`.
+These denote array types of given rank.
+- `array?`, `array2?, `array3?` , etc. are reserved words, but not `array1?` or `array0?`.
+These denote possibly-null array types of given rank.
+- `bv0`, `bv1`, `bv2`, etc. are reserved words that denote the types of
+bitvectors of given length.
+The sequence of digits after 'array' or 'bv' may not have leading zeros: 
+for example, `bv02` is an ordinary identifier.
 
 ### 2.5.2. Identifiers {#sec-identifiers}
 
-````grammar
-ident = nondigitIdChar { idchar } - charToken - reservedword
-````
-In general Dafny identifiers are sequences of ``idchar`` characters where
+In general, an `ident` token (an identifier) is a sequence of ``idchar`` characters where
 the first character is a ``nondigitIdChar``. However tokens that fit this pattern
 are not identifiers if they look like a character literal
 or a reserved word (including array or bit-vector type tokens).
 Also, `ident` tokens that begin with an `_` are not permitted as user identifiers.
 
 ### 2.5.3. Digits {#sec-digits}
-````grammar
-digits = digit {['_'] digit}
-````
 
-A sequence of decimal digits, possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
+A `digits` token is a sequence of decimal digits (`digit`), possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
 Example: `1_234_567`.
-````grammar
-hexdigits = "0x" hexdigit {['_'] hexdigit}
-````
 
-A hexadecimal constant, possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
+A `hexdigits` token denotes a hexadecimal constant, and is a sequence of hexadecimal digits (`hexdigit`)
+ possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
```

Comment:
I think we should mention the `0x` prefix here.

Reply:
Agreed. Fixed.